{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Homework 8:\n",
    "\n",
    "__Exercise 1.__ Go to prosite and find the Gamma-glutamyl phosphate reductase signature. Write a script to detect all yeast proteins that have the Gamma-glutamyl phosphate reductase signature. Your script should output a dataframe with the folowing info about the proteins: accession number, match to the first, span of the motif, and the protein's description.\n",
    "\n",
    "**NOTE:** You can just copy the file with all the yeast protein form last class in the current directory, you don't need to download it again.\n",
    "\n",
    "```\n",
    "[VA]-x(5)-A-[LIVAMTCK]-x-[HWFY]-[IM]-x(2)-[HYWNRFT]-[GSNT]-[STAG]-x(0,1)-H-[ST]-[DE]-x(1,2)-I\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'biopython'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mre\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mbiopython\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m SeqIO  \u001b[38;5;66;03m# Import SeqIO from Biopython\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Load yeast protein sequences from the FASTA file\u001b[39;00m\n\u001b[1;32m      6\u001b[0m yeast_proteins \u001b[38;5;241m=\u001b[39m SeqIO\u001b[38;5;241m.\u001b[39mparse(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtestDNA.fasta\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfasta\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'biopython'"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from biopython import SeqIO  # Import SeqIO from Biopython\n",
    "\n",
    "# Load yeast protein sequences from the FASTA file\n",
    "yeast_proteins = SeqIO.parse(\"testDNA.fasta\", \"fasta\")\n",
    "\n",
    "# Define the regular expression pattern for the Gamma-glutamyl phosphate reductase signature\n",
    "pattern = r'[VA].{5}A[LIVAMTCK].[HWFY][IM].{2}[HYWNRFT][GSNT][STAG].{0,1}H[ST][DE].{1,2}I'\n",
    "\n",
    "# Create lists to store information about matching proteins\n",
    "accession_numbers = []\n",
    "matches = []\n",
    "motif_spans = []\n",
    "descriptions = []\n",
    "\n",
    "# Iterate over each protein sequence\n",
    "for protein in yeast_proteins:\n",
    "    sequence = str(protein.seq)\n",
    "    \n",
    "    # Search for the pattern in the sequence\n",
    "    match = re.search(pattern, sequence)\n",
    "    \n",
    "    # If a match is found, extract relevant information\n",
    "    if match:\n",
    "        accession_numbers.append(protein.id)\n",
    "        matches.append(match.group())\n",
    "        motif_spans.append((match.start(), match.end()))\n",
    "        descriptions.append(protein.description)\n",
    "\n",
    "# Create a DataFrame with the extracted information\n",
    "df = pd.DataFrame({\n",
    "    'Accession Number': accession_numbers,\n",
    "    'Match to the First': matches,\n",
    "    'Span of the Motif': motif_spans,\n",
    "    'Description': descriptions\n",
    "})\n",
    "\n",
    "print(df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 2.__ Now do the same for the Hexapeptide-repeat containing-transferases signature.\n",
    "\n",
    "```\n",
    "[LIV]-[GAED]-x(2)-[STAV]-x-[LIV]-x(3)-[LIVAC]-x-[LIV]-[GAED]-x(2)-[STAVR]-x-[LIV]-[GAED]-x(2)-[STAV]-x-[LIV]-x(3)-[LIV]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from biopython import SeqIO  # Import SeqIO from Biopython\n",
    "\n",
    "# Load yeast protein sequences from the FASTA file\n",
    "yeast_proteins = SeqIO.parse(\"testDNA.fasta\", \"fasta\")\n",
    "\n",
    "# Define the regular expression pattern for the Hexapeptide-repeat containing-transferases signature\n",
    "pattern = r'[LIV][GAED].{2}[STAV].{1}[LIV].{3}[LIVAC].{1}[LIV][GAED].{2}[STAVR].{1}[LIV][GAED].{2}[STAV].{1}[LIV].{3}[LIV]'\n",
    "\n",
    "# Create lists to store information about matching proteins\n",
    "accession_numbers = []\n",
    "matches = []\n",
    "motif_spans = []\n",
    "descriptions = []\n",
    "\n",
    "# Iterate over each protein sequence\n",
    "for accession_number, sequence in yeast_proteins.items():\n",
    "    # Search for the pattern in the sequence\n",
    "    match = re.search(pattern, sequence)\n",
    "    \n",
    "    # If a match is found, extract relevant information\n",
    "    if match:\n",
    "        accession_numbers.append(accession_number)\n",
    "        matches.append(match.group())\n",
    "        motif_spans.append((match.start(), match.end()))\n",
    "        descriptions.append(\"\")\n",
    "\n",
    "# Create a DataFrame with the extracted information\n",
    "df = pd.DataFrame({\n",
    "    'Accession Number': accession_numbers,\n",
    "    'Match to the First': matches,\n",
    "    'Span of the Motif': motif_spans,\n",
    "    'Description': descriptions\n",
    "})\n",
    "\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 3.__ Now find the 14-3-3 proteins signatures. The 14-3-3 proteins seem to have multiple biological activities and play a key role in signal transduction pathways and the cell cycle. The prosite database uses two motifs to determine members of this family.\n",
    "\n",
    "Write a script to search for proteins in yeast that have both domains in either order. You should find two proteins.\n",
    "\n",
    "Your script should show a dataframe with the proteins: accession number, match to the first motif, span of the first motif, match to the second motif, span of the second motif, and the proteins description.\n",
    "\n",
    "Although your regex doesn't need to match the domains in the reverse order for it to identify both yeast proteins, I would like for you to write a regex that would be able to identify such a case for this exercise purpose.\n",
    "\n",
    "```\n",
    "[RA]-N-L-[LIV]-S-[VG]-[GA]-Y-[KN]-N-[IVA]\n",
    "```\n",
    "\n",
    "and\n",
    "\n",
    "```\n",
    "Y-K-[DE]-[SG]-T-L-I-[IML]-Q-L-[LF]-[RHC]-D-N-[LF]-T-[LS]-W-[TANS]-[SAD]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# Function to parse FASTA file\n",
    "def parse_fasta(file_path):\n",
    "    sequences = {}\n",
    "    with open(file_path, \"r\") as file:\n",
    "        current_id = None\n",
    "        current_sequence = \"\"\n",
    "        for line in file:\n",
    "            if line.startswith(\">\"):\n",
    "                if current_id:\n",
    "                    sequences[current_id] = current_sequence\n",
    "                current_id = line.strip()[1:]\n",
    "                current_sequence = \"\"\n",
    "            else:\n",
    "                current_sequence += line.strip()\n",
    "        if current_id:\n",
    "            sequences[current_id] = current_sequence\n",
    "    return sequences\n",
    "\n",
    "# Load yeast protein sequences from the FASTA file\n",
    "yeast_proteins = parse_fasta(\"testDNA.fasta\")\n",
    "\n",
    "# Define the regular expression patterns for the 14-3-3 protein motifs\n",
    "pattern1 = r'[RA]N[LIV]S[VG]A[GA]Y[KN]N[IVA]'\n",
    "pattern2 = r'YK[DE][SG]TLI[IML]QL[LF][RHC]D[NLF]T[LS]W[TANS][SAD]'\n",
    "\n",
    "# Create lists to store information about matching proteins\n",
    "accession_numbers = []\n",
    "matches1 = []\n",
    "motif_spans1 = []\n",
    "matches2 = []\n",
    "motif_spans2 = []\n",
    "descriptions = []\n",
    "\n",
    "# Iterate over each protein sequence\n",
    "for accession_number, sequence in yeast_proteins.items():\n",
    "    # Search for the first motif in the sequence\n",
    "    match1 = re.search(pattern1, sequence)\n",
    "    \n",
    "    # Search for the second motif in the sequence\n",
    "    match2 = re.search(pattern2, sequence)\n",
    "    \n",
    "    # If both motifs are found, extract relevant information\n",
    "    if match1 and match2:\n",
    "        accession_numbers.append(accession_number)\n",
    "        matches1.append(match1.group())\n",
    "        motif_spans1.append((match1.start(), match1.end()))\n",
    "        matches2.append(match2.group())\n",
    "        motif_spans2.append((match2.start(), match2.end()))\n",
    "        descriptions.append(\"\")\n",
    "\n",
    "# Create a DataFrame with the extracted information\n",
    "df = pd.DataFrame({\n",
    "    'Accession Number': accession_numbers,\n",
    "    'Match to Motif 1': matches1,\n",
    "    'Span of Motif 1': motif_spans1,\n",
    "    'Match to Motif 2': matches2,\n",
    "    'Span of Motif 2': motif_spans2,\n",
    "    'Description': descriptions\n",
    "})\n",
    "\n",
    "print(df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 4.__ Parsing and extracting data from a URL:\n",
    "\n",
    "This is form the tutorial that you should have completed.\n",
    "\n",
    "When working with files and resources over a network, you will often come across URIs and URLs which can be parsed and worked with directly. Most standard libraries will have classes to parse and construct these kind of identifiers, but if you need to match them in logs or a larger corpus of text, you can use regular expressions to pull out information from their structured format quite easily.\n",
    "\n",
    "URIs, or Uniform Resource Identifiers, are a representation of a resource that is generally composed of a scheme, host, port (optional), and resource path, respectively highlighted below.\n",
    "\n",
    "http://regexone.com:80/page\n",
    "\n",
    "The scheme describes the protocol to communicate with, the host and port describe the source of the resource, and the full path describes the location at the source for the resource.\n",
    "\n",
    "In the exercise below, try to extract the protocol, host and port of the all the resources listed in this string.\n",
    "\n",
    "```\n",
    "ftp://file_server.com:21/top_secret/life_changing_plans.pdf\n",
    "https://regexone.com/lesson/introduction#section\n",
    "file://localhost:4040/zip_file\n",
    "https://s3cur3-server.com:9999/\n",
    "market://search/angry%20birds\n",
    "```\n",
    "\n",
    "You can work interactively here: https://regexone.com/problem/extracting_url_data to find the right regular expression, then use re.finditer to create a dataframe with columns protocol, host and port for each of the matches in the string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  protocol               host  port\n",
      "0      ftp    file_server.com    21\n",
      "1    https       regexone.com  None\n",
      "2     file          localhost  4040\n",
      "3    https  s3cur3-server.com  9999\n",
      "4   market             search  None\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# Define the URLs\n",
    "urls = \"\"\"\n",
    "ftp://file_server.com:21/top_secret/life_changing_plans.pdf\n",
    "https://regexone.com/lesson/introduction#section\n",
    "file://localhost:4040/zip_file\n",
    "https://s3cur3-server.com:9999/\n",
    "market://search/angry%20birds\n",
    "\"\"\"\n",
    "\n",
    "# Define the regular expression pattern\n",
    "pattern = r'(?P<protocol>\\w+):\\/\\/(?P<host>[\\w.-]+)(?::(?P<port>\\d+))?\\/?'\n",
    "\n",
    "# Find all matches in the URLs\n",
    "matches = re.finditer(pattern, urls)\n",
    "\n",
    "# Create lists to store extracted data\n",
    "protocols = []\n",
    "hosts = []\n",
    "ports = []\n",
    "\n",
    "# Extract data and append to lists\n",
    "for match in matches:\n",
    "    protocols.append(match.group('protocol'))\n",
    "    hosts.append(match.group('host'))\n",
    "    ports.append(match.group('port'))\n",
    "\n",
    "# Create a DataFrame\n",
    "df = pd.DataFrame({\n",
    "    'protocol': protocols,\n",
    "    'host': hosts,\n",
    "    'port': ports\n",
    "})\n",
    "\n",
    "print(df)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
